cityofaustin/atd-knack
# Traffic register Automated regulation text generation

## Associated issues
https://github.com/cityofaustin/atd-data-tech/issues/8738

This PR adds code to autogenerate regulation text based on a pattern of fields. A pattern that uses `$<Knack field number>$` to mark where field values should be replaced is made for a regulation text type, and the code takes care of the value replacements.

For example, on a **ANGLEPARK** regulation type:

```
$421$ $391$ - on -  $394$ from $405$ $406$ of $403$ to $408$, $399$, $414$ side(s).
``` 
translates to 
```
Install ANGLEPARK: Angle Parking - on -  BLAKE MANOR CIR from 123 W of BRECOURT MANOR WAY to BURLESON MANOR RD, 0, EW side(s).
```

## Testing

### Test steps
You can either just review the code and trust the testing done by me and the Apps team, or you can test it out yourself by:

1. Go to https://builder.knack.com/atd/test--july-2023--traffic-register--production/records/objects/object_5. You can find the Knack Builder login in 1Password.
2. Add a row with your details so that you can log in (Non-CoA Sign-In) to the test app in the next steps. If you have issues, let me know, and I can help you out. 👍 
3. Go to https://atd.knack.com/test--july-2023--traffic-register--production#home/ and sign in (Non-CoA Sign-In)
4. From the home screen, go to **Traffic Regulations > Draft Builder**
5. Fill out the form with any details and click **Create Draft**
6. After some loading, you should see a draft regulation screen with buttons that include **Blank Install**. Click **Blank Install** and a form modal will appear.
7. Fill out every field in the form (values don't matter) and then click **Add to Draft**
8. Go to https://builder.knack.com/atd/test--july-2023--traffic-register--production/records/objects/object_22 and it should be sorted descending by **Regulation Created Timestamp**, but, if not, sort it that way.
9. You should see your record with the **Automated Regulation Text** column populated with details autogenerated by this code using what you filled into the form.

### Deployment steps
- [x] Share with apps team for feedback
- [ ] Coordinate production deployment with apps team
- [ ] Update app-specific values (app ID, api scene and view)
